src/Microsoft.AspNet.Mvc.Core/Internal/Routing/LinkGenerationMatch.cs

```diff
@@ -10,14 +13,42 @@ public struct LinkGenerationMatch
         private readonly bool _isFallbackMatch;
         private readonly AttributeRouteLinkGenerationEntry _entry;
 
-        public LinkGenerationMatch(AttributeRouteLinkGenerationEntry entry, bool isFallbackMatch)
+        public LinkGenerationMatch(
+            AttributeRouteLinkGenerationEntry entry,
+            bool isFallbackMatch,
+            VirtualPathContext pathContext)
         {
             _entry = entry;
             _isFallbackMatch = isFallbackMatch;
+            RouteValues = pathContext.Values;
+            AmbientValues = pathContext.AmbientValues;
         }
 
         public AttributeRouteLinkGenerationEntry Entry { get { return _entry; } }
 
         public bool IsFallbackMatch { get { return _isFallbackMatch; } }
+
+        public IDictionary<string, object> RouteValues { get; }
+
+        public IDictionary<string, object> AmbientValues { get; }
+
+        public int ParametersWithValues
+        {
+            get
+            {
+                var t = this;
+                return Entry?.Template?.Parameters?.Count(p => (t.RouteValues.ContainsKey(p.Name) && t.RouteValues[p.Name] != p.DefaultValue) ||
+                    (t.AmbientValues.ContainsKey(p.Name) && t.AmbientValues[p.Name] != p.DefaultValue)) ?? 0;
+            }
+        }
+
+        public int DefaultParameters
+        {
+            get
+            {
+                var t = this;
+                return Entry?.Defaults?.Count(p => t.RouteValues.ContainsKey(p.Key)) ?? 0;
```

Comment:
You want the number of parameters that have their default values. This doesn't match values, just keys.